das/s3_storage_service.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package das
+
+import (
+	"bytes"
+	"context"
+	"encoding/base32"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+
+	"github.com/offchainlabs/nitro/cmd/genericconf"
+)
+
+type S3StorageService struct {
+	s3Config   genericconf.S3Config
+	uploader   *manager.Uploader
+	downloader *manager.Downloader
+}
+
+func NewS3StorageService(s3Config genericconf.S3Config) (StorageService, error) {
+	client := s3.New(s3.Options{
+		Region:      s3Config.Region,
+		Credentials: aws.NewCredentialsCache(credentials.NewStaticCredentialsProvider(s3Config.AccessKey, s3Config.SecretKey, "")),
+	})
+	return &S3StorageService{
+		s3Config:   s3Config,
+		uploader:   manager.NewUploader(client),
+		downloader: manager.NewDownloader(client)}, nil
+}
+
+func (s3s *S3StorageService) GetByHash(ctx context.Context, key []byte) ([]byte, error) {
+	var ret []byte
+	_, err := s3s.downloader.Download(ctx, manager.NewWriteAtBuffer(ret), &s3.GetObjectInput{
+		Bucket: aws.String(s3s.s3Config.Bucket),
+		Key:    aws.String(base32.StdEncoding.EncodeToString(key)),
+	})
+
+	return ret, err
+}
+
+func (s3s *S3StorageService) PutByHash(ctx context.Context, key []byte, value []byte, timeout uint64) error {
```

Comment:
Might it make sense to have `PutByHash` queue uploads and have one or more worker threads do the actual S3 uploads?  Then `Sync` would wait until all of the queued uploads are complete.

Reply:
For now not implementing it based on the offline discussion.